From c122afdbbe28875aab21354821a0bf715eed0a3c Mon Sep 17 00:00:00 2001
From: Soumya Roy <souroy@souroy-mlt.client.nvidia.com>
Date: Mon, 31 Mar 2025 12:18:19 -0700
Subject: [PATCH] zebra: zebra crash for zapi stream Issue: If static route is
 created with a BGP route as nexthop, which recursively resolves over 512 ECMP
 v6 nexthops, zapi nexthop encode fails, as there is not enough memory
 allocated for stream. This causes assert/core dump in zebra. Right now we
 allocate fixed memory of ZEBRA_MAX_PACKET_SIZ size.

Fix:
1)Dynamically calculate required memory size for the stream
2)try to optimize memory usage

Testing:
No crash happens anymore with the fix
zebra: zebra crash for zapi stream

Issue:
If static route is created with a BGP route as nexthop, which
recursively resolves over 512 ECMP v6 nexthops, zapi nexthop encode
fails, as there is not enough memory allocated for stream. This causes
assert/core dump in zebra. Right now we allocate fixed memory
of ZEBRA_MAX_PACKET_SIZ size.

Fix:
1)Dynamically calculate required memory size for the stream
2)try to optimize memory usage

Testing:
No crash happens anymore with the fix
r1#
r1# sharp install routes 2100:cafe:: nexthop 2001:db8::1 1000
r1#

r2# conf
r2(config)# ipv6 route 2503:feca::100/128 2100:cafe::1
r2(config)# exit
r2#

Signed-off-by: Soumya Roy <souroy@nvidia.com>
---
 zebra/zebra_rnh.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/zebra/zebra_rnh.c b/zebra/zebra_rnh.c
index 83910d8a6..21e146966 100644
--- a/zebra/zebra_rnh.c
+++ b/zebra/zebra_rnh.c
@@ -1150,7 +1150,7 @@ int zebra_send_rnh_update(struct rnh *rnh, struct zserv *client,
 	re = rnh->state;
 
 	/* Get output stream. */
-	s = stream_new(ZEBRA_MAX_PACKET_SIZ);
+	s = stream_new_expandable(ZEBRA_MAX_PACKET_SIZ);
 
 	zclient_create_header(s, ZEBRA_NEXTHOP_UPDATE, vrf_id);
 
-- 
2.43.2

